influxql/ast.go

```diff
@@ -3636,61 +3656,114 @@ func evalBinaryExpr(expr *BinaryExpr, m map[string]interface{}) interface{} {
-			return lhs != rhs
-		case LT:
-			return lhs < rhs
-		case LTE:
-			return lhs <= rhs
-		case GT:
-			return lhs > rhs
-		case GTE:
-			return lhs >= rhs
-		case ADD:
-			return lhs + rhs
-		case SUB:
-			return lhs - rhs
-		case MUL:
-			return lhs * rhs
-		case DIV:
-			if rhs == 0 {
-				return float64(0)
+		switch rhs := rhs.(type) {
+		case float64:
+			switch expr.Op {
+			case EQ:
+				return lhs == rhs
+			case NEQ:
+				return lhs != rhs
+			case LT:
+				return lhs < rhs
+			case LTE:
+				return lhs <= rhs
+			case GT:
+				return lhs > rhs
+			case GTE:
+				return lhs >= rhs
+			case ADD:
+				return lhs + rhs
+			case SUB:
+				return lhs - rhs
+			case MUL:
+				return lhs * rhs
+			case DIV:
+				if rhs == 0 {
+					return float64(0)
+				}
+				return lhs / rhs
+			}
+		case int64:
+			switch expr.Op {
+			case EQ:
+				return lhs == float64(rhs)
+			case NEQ:
+				return lhs != float64(rhs)
+			case LT:
+				return lhs < float64(rhs)
+			case LTE:
+				return lhs <= float64(rhs)
+			case GT:
+				return lhs > float64(rhs)
+			case GTE:
+				return lhs >= float64(rhs)
+			case ADD:
+				return lhs + float64(rhs)
+			case SUB:
+				return lhs - float64(rhs)
+			case MUL:
+				return lhs * float64(rhs)
+			case DIV:
+				if rhs == 0 {
+					return float64(0)
+				}
+				return lhs / float64(rhs)
 			}
-			return lhs / rhs
 		}
 	case int64:
-		// we parse all number literals as float 64, so we have to convert from
-		// an interface to the float64, then cast to an int64 for comparison
-		rhsf, _ := rhs.(float64)
-		rhs := int64(rhsf)
-		switch expr.Op {
-		case EQ:
-			return lhs == rhs
-		case NEQ:
-			return lhs != rhs
-		case LT:
-			return lhs < rhs
-		case LTE:
-			return lhs <= rhs
-		case GT:
-			return lhs > rhs
-		case GTE:
-			return lhs >= rhs
-		case ADD:
-			return lhs + rhs
-		case SUB:
-			return lhs - rhs
-		case MUL:
-			return lhs * rhs
-		case DIV:
-			if rhs == 0 {
-				return int64(0)
+		switch rhs := rhs.(type) {
+		case float64:
+			switch expr.Op {
+			case EQ:
+				return float64(lhs) == rhs
+			case NEQ:
+				return float64(lhs) != rhs
+			case LT:
+				return float64(lhs) < rhs
+			case LTE:
+				return float64(lhs) <= rhs
+			case GT:
+				return float64(lhs) > rhs
+			case GTE:
+				return float64(lhs) >= rhs
+			case ADD:
+				return float64(lhs) + rhs
+			case SUB:
+				return float64(lhs) - rhs
+			case MUL:
+				return float64(lhs) * rhs
+			case DIV:
+				if rhs == 0 {
+					return float64(0)
+				}
+				return float64(lhs) / rhs
+			}
+		case int64:
+			switch expr.Op {
+			case EQ:
+				return lhs == rhs
+			case NEQ:
+				return lhs != rhs
+			case LT:
+				return lhs < rhs
+			case LTE:
+				return lhs <= rhs
+			case GT:
+				return lhs > rhs
+			case GTE:
+				return lhs >= rhs
+			case ADD:
+				return lhs + rhs
+			case SUB:
+				return lhs - rhs
+			case MUL:
+				return lhs * rhs
+			case DIV:
+				if rhs == 0 {
+					return float64(0)
+				}
+				return float64(lhs) / float64(rhs)
```

Comment:
The first one can be changed, but this one needs to be two separate switches because the first is a float/float comparison and the second is an integer/integer comparison. I can't have the types be either a float or an integer, so I need two different switches.